### PR TITLE
Block new requests to non busy projects during maintenance mode

### PR DESCRIPTION
**Pull request by inaod568**
> **Before submitting the PR, please go through the sections below and fill in what you can. If there are any items that are irrelevant for the current PR, remove the row. If a relevant option is missing, please add it as an item and add a PR comment informing that the new option should be included into this template.**

> **All _relevant_ items should be ticked before the PR is merged**

# Description

Was hoping for a "before_request" thing as in `dds_web/__init__.py` but there is none for Flask-Restful Resources. Also tested adding the check to `get_user_roles_common` which gets the user role, but that one is not run if the resource allows all roles. Now a new function is called in the before_request to check if maintenance is ongoing, and in that case if the project is busy and trying to access an approved endpoint.

We want to be able to stop new incoming requests when we set the DDS to maintenance mode, but not cancel ongoing ones.

- [x] Add a summary of the changes and the related issue
- [x] Add motivation and context regarding why the change is needed
- [x] Fixes DDS-1347

## Type of change

- [x] New feature (non-breaking)

# Checklist:

## General

- [x] Code change
  - [x] Self-review of code done
  - [x] Comments added, particularly in hard-to-understand areas

## Repository / Releases

- [x] Blocking PRs have been merged
- [x] Rebase / update of branch done
- [x] Product Owner / Scrum Master

## Checks

- [x] Formatting: Black & Prettier checks pass
- [x] CodeQL passes
- [x] Tests
  - [x] I have added tests for the new code
  - [x] The tests pass
- [x] Trivy / Snyk:
  - [x] There are no new security alerts

<a href="https://gitpod.io/#https://github.com/ScilifelabDataCentre/dds_web/pull/1289"></a>

### CODE REVIEW

**Comment by inaod568** [comment]
Have been trying to get a page with the 503: Service Unavailable working, by aborting in the `before_request`  but it's not rendering properly. Pushing to PR now, try it out and then compare to how it should look by adding e.g. "flask.abort(503)" in def home():

File: dds_web/api/dds_decorators.py [hunk 1]
```diff
@@ -22,10 +22,13 @@
     DatabaseError,
     DDSArgumentError,
     LoggedHTTPException,
+    MaintenanceOngoingException,
```
Comment (inaod568) [inline dds_web/api/dds_decorators.py:25]:
```suggestion
```

File: dds_web/api/dds_decorators.py [hunk 2]
```diff
@@ -22,10 +22,13 @@
...
     MissingJsonError,
     S3ConnectionError,
 )
-from dds_web.utils import get_username_or_request_ip
+from dds_web.utils import collect_project, get_username_or_request_ip
+from dds_web.database import models
+from dds_web import auth
```
Comment (inaod568) [inline dds_web/api/dds_decorators.py:31]:
```suggestion
from dds_web.utils import get_username_or_request_ip
```

File: dds_web/errors.py [hunk 3]
```diff
@@ -423,3 +423,13 @@ def __init__(
     ):
         super().__init__(message)
         general_logger.warning(message)
+
+
+class MaintenanceOngoingException(LoggedHTTPException):
+
+    code = http.HTTPStatus.FORBIDDEN
```
Comment (inaod568) [inline dds_web/errors.py:430]:
```suggestion
    code = http.HTTPStatus.SERVICE_UNAVAILABLE
```

File: dds_web/security/auth.py [hunk 4]
```diff
@@ -18,7 +18,13 @@
...
 # Own modules
 from dds_web import basic_auth, auth, mail
-from dds_web.errors import AuthenticationError, AccessDeniedError, InviteError, TokenMissingError
+from dds_web.errors import (
+    AuthenticationError,
+    AccessDeniedError,
+    InviteError,
+    MaintenanceOngoingException,
+    TokenMissingError,
+)
```
Comment (inaod568) [inline dds_web/security/auth.py:27]:
```suggestion
from dds_web.errors import (
    AuthenticationError,
    AccessDeniedError,
    InviteError,
    TokenMissingError,
)
```

File: dds_web/utils.py [hunk 5]
```diff
@@ -18,6 +18,7 @@
 import flask
 from dds_web.errors import (
     AccessDeniedError,
+    MaintenanceOngoingException,
```
Comment (inaod568) [inline dds_web/utils.py:21]:
```suggestion
```

**Comment by inaod568** [comment]
@valyo Have made changes, have a look and tell me what you think

**Review by valyo — APPROVED** [review]
The only comment I have is that all blocked commands produce a stack trace in the dds_web logs:
```
dds_backend      | [2022-09-29 11:38:22,387] errors [WARNING] Maintenance of DDS is ongoing.
dds_backend      | [2022-09-29 11:38:22,388] app [ERROR] Exception on /api/v1/files/list [GET]
dds_backend      | Traceback (most recent call last):
dds_backend      |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1514, in full_dispatch_request
dds_backend      |     rv = self.preprocess_request()
dds_backend      |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1857, in preprocess_request
dds_backend      |     rv = self.ensure_sync(before_func)()
dds_backend      |   File "/code/dds_web/__init__.py", line 196, in prepare
dds_backend      |     block_if_maintenance()
dds_backend      |   File "/code/dds_web/utils.py", line 600, in block_if_maintenance
dds_backend      |     raise MaintenanceOngoingException()
dds_backend      | dds_web.errors.MaintenanceOngoingException: 503 Service Unavailable: Maintenance of DDS is ongoing.
```
not sure how this can be avoided and if it is so important

**Comment by inaod568** [comment]
> The only comment I have is that all blocked commands produce a stack trace in the dds_web logs:
> 
> ```
> dds_backend      | [2022-09-29 11:38:22,387] errors [WARNING] Maintenance of DDS is ongoing.
> dds_backend      | [2022-09-29 11:38:22,388] app [ERROR] Exception on /api/v1/files/list [GET]
> dds_backend      | Traceback (most recent call last):
> dds_backend      |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1514, in full_dispatch_request
> dds_backend      |     rv = self.preprocess_request()
> dds_backend      |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1857, in preprocess_request
> dds_backend      |     rv = self.ensure_sync(before_func)()
> dds_backend      |   File "/code/dds_web/__init__.py", line 196, in prepare
> dds_backend      |     block_if_maintenance()
> dds_backend      |   File "/code/dds_web/utils.py", line 600, in block_if_maintenance
> dds_backend      |     raise MaintenanceOngoingException()
> dds_backend      | dds_web.errors.MaintenanceOngoingException: 503 Service Unavailable: Maintenance of DDS is ongoing.
> ```
> 
> not sure how this can be avoided and if it is so important

@valyo Hmm, yeah I think this is in the config of the MaintenanceOngoingException, it's currently logging to the general logger as well.

File: tests/test_utils.py [hunk 6]
```diff
@@ -19,6 +19,9 @@
 from flask.testing import FlaskClient
 import requests_mock
 import werkzeug
+from tests import DDSEndpoint, DEFAULT_HEADER, UserAuth, USER_CREDENTIALS
+import http
+import time
```
Comment (inaod568) [inline tests/test_utils.py:24]:
```suggestion
```

File: dds_web/errors.py [hunk 7]
```diff
@@ -423,3 +423,13 @@ def __init__(
...
+
+
+class MaintenanceOngoingException(LoggedHTTPException):
+
+    code = http.HTTPStatus.SERVICE_UNAVAILABLE
+
+    def __init__(self, message="Maintenance of DDS is ongoing."):
+        """Inform that maintenance is ongoing."""
+        super().__init__(message)
+        general_logger.warning(message)
```
Comment (inaod568) [inline dds_web/errors.py:435]:
```suggestion
```

**Comment by inaod568** [comment]
@valyo I know you've approved this previously but since there were changes and I've added tests a new approval would be good when you've had a look etc.

File: dds_web/utils.py [hunk 8]
```diff
@@ -571,3 +572,49 @@ def calculate_version_period_usage(version):
...
+            raise MaintenanceOngoingException()
+        else:
+            # Request accepted during maintenance but...
+            req_args = flask.request.args
+            if current_endpoint in project_required_endpoints:
+                # Request requires a project
+                if not (req_args and (project_id := req_args.get("project"))):
+                    raise MaintenanceOngoingException()
+
+                # Request requires an busy project
```
Comment (valyo) [inline dds_web/utils.py:616]: # Request requires an busy project
should be
# Request requires a busy project

**Review by valyo — APPROVED** [review]
My manual tests worked as expected. 
I added comments about some dicstrings, but not all the "wrong" one. Probably not very important, but just because it can be misleading for someone in the future.

**Pull request merged**